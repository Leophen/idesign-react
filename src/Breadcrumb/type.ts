export interface BreadcrumbProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 单项最大宽度，超出后会以省略号形式呈现
   */
  maxItemWidth?: string | number;
  /**
   * 自定义分隔符
   */
  separator?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
}

export interface BreadcrumbItemProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 是否禁用当前项点击
   */
  disabled?: boolean;
  /**
   * 单项最大宽度，超出后会以省略号形式呈现
   */
  maxItemWidth?: string | number;
  /**
   * 最大宽度，超出后会以省略号形式呈现。优先级高于 maxItemWidth
   */
  maxWidth?: string | number;
  /**
   * 自定义分隔符
   */
  separator?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
}
