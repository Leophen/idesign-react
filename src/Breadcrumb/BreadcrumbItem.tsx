import React, { useMemo } from 'react';
import classNames from 'classnames';
import Icon from '../Icon';
import './index.scss';
import { BreadcrumbItemProps } from './type';

const BreadcrumbItem: React.FC<BreadcrumbItemProps> = (props) => {
  const {
    children = '',
    className,
    disabled,
    maxItemWidth,
    maxWidth,
    separator = <Icon name="ArrowRight" size={16} />,
    style,
    ...restProps
  } = props;

  // 限制最大宽度
  const computedMaxWidth = useMemo(() => {
    let curMaxWidth;
    if (maxWidth) {
      curMaxWidth = typeof maxWidth === 'number' ? `${maxWidth}px` : String(maxWidth);
    }
    let curMaxItemWidth;
    if (maxItemWidth) {
      curMaxItemWidth =
        typeof maxItemWidth === 'number' ? `${maxItemWidth}px` : String(maxItemWidth);
    }
    return {
      maxWidth: curMaxWidth || curMaxItemWidth || '200px',
    };
  }, [maxItemWidth, maxWidth]);

  return (
    <div
      className={classNames(
        'i-breadcrumb__item',
        disabled && 'i-breadcrumb-is-disabled',
        className,
      )}
      style={{ ...style }}
      {...restProps}
    >
      <span className="i-breadcrumb__inner" style={computedMaxWidth}>
        {children}
      </span>
      <span className="i-breadcrumb__separator">{separator}</span>
    </div>
  );
};

export default BreadcrumbItem;
