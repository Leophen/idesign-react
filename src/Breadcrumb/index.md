---
nav:
  title: 组件
  path: /components
group:
  title: 导航组件
  order: 2
order: 3
---

# Breadcrumb 面包屑

显示当前页面在系统层级结构的位置，并能返回之前任意层级的页面。

## 基本用法

使用 `<Breadcrumb>` 包裹面包屑的每一项 `<Breadcrumb.Item>`。

```tsx
import React from 'react';
import { Breadcrumb } from 'idesign-react';

const App = () => {
  return (
    <Breadcrumb>
      <Breadcrumb.Item>页面1</Breadcrumb.Item>
      <Breadcrumb.Item>页面2222222222222222222222222222222</Breadcrumb.Item>
      <Breadcrumb.Item>页面3333333333333333333333333333333</Breadcrumb.Item>
    </Breadcrumb>
  );
};

export default App;
```

## 禁用状态

`<Breadcrumb.Item>` 可以使用 `disabled` 属性控制单项为禁用状态。

```tsx
import React from 'react';
import { Breadcrumb } from 'idesign-react';

const App = () => {
  return (
    <Breadcrumb>
      <Breadcrumb.Item>页面1</Breadcrumb.Item>
      <Breadcrumb.Item disabled>页面2222222222222222222222222222222</Breadcrumb.Item>
      <Breadcrumb.Item>页面3333333333333333333333333333333</Breadcrumb.Item>
    </Breadcrumb>
  );
};

export default App;
```

## 自定义单项最大宽度

面包屑每一项的最大宽度默认为 `200px`；

`<Breadcrumb>` 可以使用 `maxItemWidth` 属性来全局控制面包屑每一项的最大宽度，内容超出宽度会以省略号形式呈现；

`<Breadcrumb.Item>` 可以使用 `maxWidth` 属性来控制单项的最大宽度，内容超出宽度会以省略号形式呈现，优先级高于 `maxItemWidth`。

```tsx
import React from 'react';
import { Breadcrumb } from 'idesign-react';

const App = () => {
  return (
    <>
      <Breadcrumb maxItemWidth="120px">
        <Breadcrumb.Item>页面1</Breadcrumb.Item>
        <Breadcrumb.Item>页面2222222222222222222222222222222</Breadcrumb.Item>
        <Breadcrumb.Item>页面3333333333333333333333333333333</Breadcrumb.Item>
      </Breadcrumb>

      <br />

      <Breadcrumb maxItemWidth="120px">
        <Breadcrumb.Item>页面1</Breadcrumb.Item>
        <Breadcrumb.Item maxWidth="80px">页面2222222222222222222222222222222</Breadcrumb.Item>
        <Breadcrumb.Item>页面3333333333333333333333333333333</Breadcrumb.Item>
      </Breadcrumb>

      <br />

      <Breadcrumb maxItemWidth={120}>
        <Breadcrumb.Item>页面1</Breadcrumb.Item>
        <Breadcrumb.Item maxWidth="80px">页面2222222222222222222222222222222</Breadcrumb.Item>
        <Breadcrumb.Item>页面3333333333333333333333333333333</Breadcrumb.Item>
      </Breadcrumb>

      <br />

      <Breadcrumb maxItemWidth={120}>
        <Breadcrumb.Item>页面1</Breadcrumb.Item>
        <Breadcrumb.Item maxWidth={80}>页面2222222222222222222222222222222</Breadcrumb.Item>
        <Breadcrumb.Item>页面3333333333333333333333333333333</Breadcrumb.Item>
      </Breadcrumb>
    </>
  );
};

export default App;
```

## 自定义分隔符

`<Breadcrumb>` 可以使用 `separator` 属性来自定义面包屑每一项的分隔符；

```tsx
import React from 'react';
import { Breadcrumb } from 'idesign-react';

const App = () => {
  return (
    <Breadcrumb separator="👉">
      <Breadcrumb.Item>页面1</Breadcrumb.Item>
      <Breadcrumb.Item>页面2222222222222222222222222222222</Breadcrumb.Item>
      <Breadcrumb.Item>页面3333333333333333333333333333333</Breadcrumb.Item>
    </Breadcrumb>
  );
};

export default App;
```

## 路由跳转

可通过包裹 `<a>` 标签或 [React Router](https://reactrouter.com/) 的 `<Link>` 来实现路由跳转。

```tsx
import React from 'react';
import { Breadcrumb } from 'idesign-react';
import { Link } from 'react-router-dom';

const App = () => {
  return (
    <Breadcrumb>
      <Breadcrumb.Item>
        <a href="/components/breadcrumb">跳转方式一</a>
      </Breadcrumb.Item>
      <Breadcrumb.Item>
        <Link to="/components/breadcrumb">跳转方式二</Link>
      </Breadcrumb.Item>
    </Breadcrumb>
  );
};

export default App;
```

## Breadcrumb API

| 属性         | 说明                                     | 类型              | 默认值  |
| ------------ | ---------------------------------------- | ----------------- | ------- |
| children     | 按钮内容                                 | `ReactNode`       | `--`      |
| className    | 类名                                     | `string`          | `--`      |
| maxItemWidth | 各项的最大宽度，超出后会以省略号形式呈现 | `string / number` | `200px` |
| separator    | 自定义分隔符                             | `ReactNode`       | `--`      |
| style        | 自定义样式                               | `CSSProperties`   | `--`      |

## BreadcrumbItem API

| 属性      | 说明                                  | 类型              | 默认值  |
| --------- | ------------------------------------- | ----------------- | ------- |
| children  | 按钮内容                              | `ReactNode`       | `--`      |
| className | 类名                                  | `string`          | `--`      |
| style     | 自定义样式                            | `CSSProperties`   | `--`      |
| disabled  | 是否禁用当前项点击                    | `boolean`         | `false` |
| maxWidth  | 单项最大宽度，优先级高于 maxItemWidth | `string / number` | `--`      |
