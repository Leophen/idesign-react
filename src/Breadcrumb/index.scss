.i-breadcrumb {
  display: flex;
  align-items: center;
}

.i-breadcrumb__item {
  display: flex;
  align-items: center;
  &:last-of-type {
    .i-breadcrumb__inner,
    .i-breadcrumb__inner a {
      color: var(--i-font2);
      &:hover {
        color: var(--i-primary-txt);
        transition: color 0.2s ease-out;
      }
    }
    .i-breadcrumb__separator {
      display: none;
    }
  }
}

.i-breadcrumb__inner,
.i-breadcrumb__inner a {
  display: inline-block;
  overflow: hidden;
  color: var(--i-font-4);
  white-space: nowrap;
  text-overflow: ellipsis;
  word-break: break-all;
  cursor: pointer;
  &:hover {
    color: var(--i-primary-txt);
    transition: color 0.2s ease-out;
  }
}

.i-breadcrumb__separator {
  display: flex;
  align-items: center;
  margin: 0 4px;
  .i-icon {
    color: var(--i-font-4);
  }
}

.i-breadcrumb-is-disabled {
  .i-breadcrumb__inner,
  .i-breadcrumb__inner a {
    color: var(--i-font-disabled);
    cursor: not-allowed;
    &:hover {
      color: var(--i-font-disabled);
    }
  }
}
