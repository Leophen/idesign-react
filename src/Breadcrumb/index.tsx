import React from 'react';
import classNames from 'classnames';
import './index.scss';
import BreadcrumbItem from './BreadcrumbItem'
import { BreadcrumbProps } from './type';

const Breadcrumb: React.FC<BreadcrumbProps> & { Item: React.ElementType } = (props) => {
  const {
    children = '',
    className,
    maxItemWidth,
    separator,
    style,
    ...restProps
  } = props;

  return (
    <div
      className={classNames(
        'i-breadcrumb',
        className
      )}
      style={{ ...style }}
      {...restProps}
    >
      {/* 将 separator 分隔符及 maxItemWidth 最大单项宽度传入 children */}
      {React.Children.map(children, (child) => {
        if (!React.isValidElement(child)) {
          return null;
        }
        const childProps = {
          separator,
          maxItemWidth,
        };
        return React.cloneElement(child, childProps);
      })}
    </div>
  );
};

Breadcrumb.Item = BreadcrumbItem;

BreadcrumbItem.displayName = 'BreadcrumbItem';
Breadcrumb.displayName = 'Breadcrumb';

export default Breadcrumb;
