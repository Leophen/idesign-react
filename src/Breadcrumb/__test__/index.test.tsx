import { render } from '@testing-library/react';
import Breadcrumb from '../index';

describe('Breadcrumb ç»„ä»¶æµ‹è¯•', () => {
  it('create', () => {
    const el = (
      <Breadcrumb>
        <Breadcrumb.Item>item1</Breadcrumb.Item>
        <Breadcrumb.Item>item2</Breadcrumb.Item>
        <Breadcrumb.Item>item3</Breadcrumb.Item>
      </Breadcrumb>
    );
    const { asFragment } = render(el);
    expect(asFragment()).toMatchSnapshot();
  });

  it('breadcrumbItem children', () => {
    const el = (
      <Breadcrumb>
        <Breadcrumb.Item>item</Breadcrumb.Item>
      </Breadcrumb>
    );
    const { container } = render(el);
    expect(container.firstChild.firstChild).toHaveTextContent('item');
  });

  it('breadcrumbItem disabled', () => {
    const el = (
      <Breadcrumb>
        <Breadcrumb.Item disabled>item</Breadcrumb.Item>
      </Breadcrumb>
    );
    const { container } = render(el);
    expect(container.firstChild.firstChild).toHaveClass('i-breadcrumb-is-disabled');
  });

  it('max width', () => {
    const el = (
      <Breadcrumb maxItemWidth="120px">
        <Breadcrumb.Item>item</Breadcrumb.Item>
      </Breadcrumb>
    );
    const { container } = render(el);
    expect(container.firstChild.firstChild.firstChild).toHaveStyle('max-width: 120px');
  });

  it('separator', () => {
    const el = (
      <Breadcrumb separator="ðŸ‘‰">
        <Breadcrumb.Item>item</Breadcrumb.Item>
      </Breadcrumb>
    );
    const { container } = render(el);
    expect(container.firstChild.firstChild.lastChild).toHaveTextContent('ðŸ‘‰');
  });
});

