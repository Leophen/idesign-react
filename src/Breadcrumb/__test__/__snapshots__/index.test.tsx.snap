// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-breadcrumb"
  >
    <div
      class="i-breadcrumb__item"
    >
      <span
        class="i-breadcrumb__inner"
        style="max-width: 200px;"
      >
        item1
      </span>
      <span
        class="i-breadcrumb__separator"
      >
        <i
          class="i-icon icon-ArrowRight"
          style="font-size: 16px;"
        />
      </span>
    </div>
    <div
      class="i-breadcrumb__item"
    >
      <span
        class="i-breadcrumb__inner"
        style="max-width: 200px;"
      >
        item2
      </span>
      <span
        class="i-breadcrumb__separator"
      >
        <i
          class="i-icon icon-ArrowRight"
          style="font-size: 16px;"
        />
      </span>
    </div>
    <div
      class="i-breadcrumb__item"
    >
      <span
        class="i-breadcrumb__inner"
        style="max-width: 200px;"
      >
        item3
      </span>
      <span
        class="i-breadcrumb__separator"
      >
        <i
          class="i-icon icon-ArrowRight"
          style="font-size: 16px;"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
